### ISSUE TITLE
Need to useful examples of things people might do in real life.

### ISSUE THREAD

**Issue by rsdoiel** [rsdoiel/sqlish#42]

**Comment by yawnt** [comment]
i'm willing to rewrite https://github.com/yawnt/hater using sqlish.. perhaps we could collaborate to make sqlish a real world example?

**Comment by rsdoiel** [comment]
I'd be glad to.  I've been using the current Sqlish in some work things, so
far it has been helpful.

Thanks,

Robert

**Comment by yawnt** [comment]
ok awesome, @stoke and i started a branch (rewrite) in hater and i'll let you know if we find issues.. my main concern atm is variable sanitization.. what's the status on that?

**Comment by rsdoiel** [comment]
Well I'm still not 100% happy with it. I need to build some more test cases
based on real world usage. I believe I broke SQLite quoting when I did the
last code refactoring (though I haven't written test cases to prove it
yet).  I'm coming around to the belief that your original proposal to
define your tables first is a requirement to do the job right and keep it
simple. I have some of that implemented but remain split with my original
approach.

What I am hoping to have time to sort out is something really simple like
this -

sqlish.schemas({"mytable", {
      id: Number,
      name: String,
      modified: Date,
     },
     "mytable2", {
       ....
    }
);

I like using the JS primitives because that's what I really worried about
and I think they will map into most SQL. I would still allow
createTable to be more explicit.

thanks,

Robert

**Comment by yawnt** [comment]
Uhm i see what you mean.. the question is that there are a lot more types in SQL than in js.. how would you make a distinction between Date and Datetime for example?

**Comment by rsdoiel** [comment]
Yes, that's my tricky part to think through. Right now I'm thinking I could
fake it by creating an Alias for DATETIME, DATE, TIME that correspond to
the JavaScript object but allow for more controlled generation of the SQL
statement.  In terms of preventing injection they would have to be a child
of Date but I would have enough meta data to know that resolve to something
specific (e.g. a MySQL Date, Datetime, or Time).  Still thinking it through.

Thanks,

Robert

**Comment by yawnt** [comment]
i see.. when do you think you'll be able to work this out? or if you have any sketch i or @stoke could implement that and submit a PR ;)

// yawnt

**Comment by rsdoiel** [comment]
I in meetings most of today and tomorrow so I will have to do it at night.
Let's say Friday evening. Is that soon enough?

Thanks,

Robert

**Comment by yawnt** [comment]
take your time, no hurry.. it's just so that i can schedule my rewrite :)

**Comment by rsdoiel** [comment]
I'll shoot for something end of weekend then.  As always thank you for the fixes. Does @stoke need to be added as a collaborators?

**Comment by yawnt** [comment]
i don't know if he wants to work here or in a fork, just add him.. he'll decide what to do :)

**Comment by stoke** [comment]
Nah, a fork is fine :D

**Comment by rsdoiel** [comment]
@yawnt. Ok, my ducks are beginning to line up in a row for version 0.0.9 (which I hope to release soon, maybe shortly after next weekend).  0.0.9 adds a safeExpr() method which is inline with my original approach to defend against injection. It is not the only defense against injection though. I think the schema approach is more specific and therefore less error prone but I've found I still need to be able to have a lazy option available (this is based on my own usage in some work related projects where I'm using Sqlish for to help in data migrations from MySQL to MongoDB).

If a schema has been defined (e.g. say by Hater), then Sqlish will rely on that to vet the inputs used to render the final SQL.   I'll have the schema code sorted by the 0.0.9 release but I'm not going to have a chance to really look at again until next weekend. I want it really simple to use and I don't think I'm there yet architecturally.

I still have to solve the escape issue for SQLite versus everyone else.  If I'm not happy with my approach by the time I sort out the schema definition than I drop support for SQLite in the next 0.0.9 release. Since I'm claiming I'm defending against injection I don't want to release a version that only sorta handles quoting right.

Anyway, before 0.0.9 is final I need to do a code review and cleanup. My quick and dirty pass at safeExpr() is very crude. I'm seeing lots of crufty stuff that makes me cringe.

All the best,

Robert

**Comment by yawnt** [comment]
looks good, thanks for taking the time to look into this..

lemme know if i can help in any way

// yawnt

**Comment by rsdoiel** [comment]
When I'm ready I'll need at least one more pair of eyes for the code
review. It'll also be helpful to get more test coverage for injection. If
you have some good examples we can turn into a test battery that'd be
appreciated. Otherwise I'll to do a net search and see what I can find.
 all the best,

Robert

**Comment by rsdoiel** [comment]
Good morning,

I'm using Sqlish in a work project and have made some small tweaks in the
0_0_9 branch. It really is turning out quiet useful. I believe ad-hoc
validation may actually be working because it has caught some unintended
errors I made in the program using the module. I think I should take a
closer look at some SQL injecting test libraries to make sure I have
reasonable coverage. I'll be away at conference next week so my guess is I
will not push 0_0_9 to master until after to Thanksgiving.

All the best,

Robert

**Comment by yawnt** [comment]
Ok, that's cool, i also have conferences to attend this weekend (nodejsconf.it) so it's fine :P 

// yawnt

**Comment by rsdoiel** [comment]
That should be very fun.  Let me know the interesting things when you get
back.  I'm heading to YUIConf.  Going to YUI is a little funny because I
don't work in a YUI shop (the front end folks here are jQuery people) but
the time I went a before I came back with allot of ideas. They were first
folks I met in person doing Node "for development only" at the time. Must
have been two year ago at least.  Anyhow the Bay Area is nice drive back
along the coast to South California is quite beautiful.

All the best,

Robert
